### PR TITLE
cmake: Check for GNU compiler, not Linux, to set -D_GNU_SOURCE

### PR DESCRIPTION
**Pull request by jamessan**
The Debian hurd-i386 [build] failed (partly) due to -D_GNU_SOURCE not be
defined:

    [215/286] /usr/bin/cc -DINCLUDE_GENERATED_DECLARATIONS -Iconfig -I../src -Isrc/nvim/auto -Iinclude -I/usr/include/luajit-2.1 -g -O2 -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -DDISABLE_LOG -Wdate-time -D_FORTIFY_SOURCE=2 -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -O2 -g  -DMIN_LOG_LEVEL=3 -Og -g   -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -Wno-array-bounds -MD -MT src/nvim/CMakeFiles/nvim.dir/os/pty_process_unix.c.o -MF src/nvim/CMakeFiles/nvim.dir/os/pty_process_unix.c.o.d -o src/nvim/CMakeFiles/nvim.dir/os/pty_process_unix.c.o   -c ../src/nvim/os/pty_process_unix.c
    ../src/nvim/os/pty_process_unix.c: In function 'pty_process_tty_name':
    ../src/nvim/os/pty_process_unix.c:121:10: warning: implicit declaration of function 'ptsname'; did you mean 'ttyname'? [-Wimplicit-function-declaration]
       return ptsname(ptyproc->tty_fd);

Hurd is obviously not Linux, but it is using a GNU compiler and glibc so
it needs -D_GNU_SOURCE for the ptsname() definition to be visible.

[build]: https://buildd.debian.org/status/fetch.php?pkg=neovim&arch=hurd-i386&ver=0.3.0-2&stamp=1528981349&raw=0

### CODE REVIEW

**Comment by justinmk** [comment]
Does this still work for clang?

**Comment by jamessan** [comment]
> Does this still work for clang?

Apparently not. :) I'll take this opportunity to cleanup some other stuff in the CMake files when I fix that.

**Comment by jamessan** [comment]
There's one more commit (raising minimum version of CMake to 2.8.12) I'll get to later today.

File: CMakeLists.txt [hunk 1]
```diff
@@ -2,13 +2,9 @@
 #   intro: https://codingnest.com/basic-cmake/
 #   best practices (3.0+): https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1
 
-cmake_minimum_required(VERSION 2.8.7)
+cmake_minimum_required(VERSION 2.8.12)
 project(nvim C)
 
-if(POLICY CMP0059)
-  cmake_policy(SET CMP0059 OLD)  # Needed until cmake 2.8.12. #4389
```
Comment (justinmk) [inline CMakeLists.txt]: there might be some issue(s) that can be closed.

File: CMakeLists.txt [hunk 2]
```diff
@@ -2,13 +2,9 @@
 #   intro: https://codingnest.com/basic-cmake/
 #   best practices (3.0+): https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1
 
-cmake_minimum_required(VERSION 2.8.7)
+cmake_minimum_required(VERSION 2.8.12)
```
Comment (justinmk) [inline CMakeLists.txt]: not sure if it matters, but this also shows up in:

```
third-party/CMakeLists.txt|2| cmake_minimum_required (VERSION 2.8.7)
third-party/cmake/GettextCMakeLists.txt|1| cmake_minimum_required(VERSION 2.8.7)
third-party/cmake/GperfCMakeLists.txt|1| cmake_minimum_required(VERSION 2.8.7)
third-party/cmake/LibiconvCMakeLists.txt|1| cmake_minimum_required(VERSION 2.8.7)
third-party/cmake/LibuvCMakeLists.txt|1| cmake_minimum_required(VERSION 2.8.7)
third-party/cmake/Libvterm-tbl2inc_c.cmake|1| cmake_minimum_required(VERSION 2.8.7)
third-party/cmake/LibvtermCMakeLists.txt|1| cmake_minimum_required(VERSION 2.8.11)
third-party/cmake/UnibiliumCMakeLists.txt|1| cmake_minimum_required(VERSION 2.8.7)
third-party/cmake/libtermkeyCMakeLists.txt|1| cmake_minimum_required(VERSION 2.8.7)
```

**Comment by justinmk** [comment]
@jszakmeister Could you sanity-check this, since some of it is your original work?

File: CMakeLists.txt (hunk 2 above)
Comment (jamessan) [reply CMakeLists.txt]: Ah, right.  I missed third-party somehow. :)

**Comment by jszakmeister** [comment]
Is there a good way to check if `-lc` is actually glibc or do other C libraries also require `_GNU_SOURCE`?  It looks like this might be a reasonable way to test for glibc: https://sourceforge.net/p/predef/wiki/Libraries/#gnu-glibc

**Comment by justinmk** [comment]
quickbuild BSD env is complaining:

```
11:26:14,889 WARN  - CMakeFiles/nvim-test.dir/os/env.c.o: In function `os_getenvname_at_index':
11:26:14,890 WARN  - /usr/home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/os/env.c:(.text+0x7c): undefined reference to `environ'
11:26:14,890 WARN  - CMakeFiles/nvim-test.dir/os/env.c.o: In function `get_env_name':
11:26:14,890 WARN  - /usr/home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/os/env.c:(.text+0x10b1): undefined reference to `environ'
11:26:14,899 WARN  - ../../../../.deps/usr/lib/libuv.a(libuv_la-process.o): In function `uv__process_child_init':
11:26:14,899 WARN  - /usr/home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/.deps/build/src/libuv/src/unix/process.c:376: undefined reference to `environ'
11:26:14,927 WARN  - cc: error: linker command failed with exit code 1 (use -v to see invocation)
11:26:14,929 WARN  - gmake[3]: *** [src/nvim/CMakeFiles/nvim-test.dir/build.make:4381: lib/libnvim-test.so] Error 1
11:26:14,929 INFO  - gmake[3]: Leaving directory '/usr/home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/build/Release'
11:26:14,929 WARN  - gmake[2]: *** [CMakeFiles/Makefile2:3853: src/nvim/CMakeFiles/nvim-test.dir/all] Error 2
11:26:14,929 INFO  - gmake[2]: Leaving directory '/usr/home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/build/Release'
11:26:14,930 WARN  - gmake[1]: *** [CMakeFiles/Makefile2:341: CMakeFiles/unittest-prereqs.dir/rule] Error 2
11:26:14,930 INFO  - gmake[1]: Leaving directory '/usr/home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/build/Release'
11:26:14,930 WARN  - gmake: *** [Makefile:288: unittest-prereqs] Error 2
11:26:14,932 INFO  - Executing post-execute action...
11:26:14,932 ERROR - Step 'master>buildall>build-node?testNode=freebsd-64>build-and-run-tests>build-and-run-tests-parameterized?buildType=Release>configure-neovim-and-build-nvim' is failed: Failed to run command: mkdir -p build/Release && cd build/Release && cmake -G "Unix Makefiles" -DBUSTED_OUTPUT_TYPE=TAP -DCMAKE_BUILD_TYPE=Release -DTRAVIS_CI_BUILD=ON ../.. && gmake VERBOSE=1 nvim unittest-prereqs functionaltest-prereqs
Command return code: 2
Command error output: CMakeFiles/nvim-test.dir/os/env.c.o: In function `os_getenvname_at_index':
/usr/home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/os/env.c:(.text+0x7c): undefined reference to `environ'
CMakeFiles/nvim-test.dir/os/env.c.o: In function `get_env_name':
/usr/home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/os/env.c:(.text+0x10b1): undefined reference to `environ'
../../../../.deps/usr/lib/libuv.a(libuv_la-process.o): In function `uv__process_child_init':
/usr/home/quickbuild/buildagent/workspace/root/n...
11:26:15,092 INFO  - Checking step execute condition...
```

**Comment by jamessan** [comment]
> Is there a good way to check if -lc is actually glibc or do other C libraries also require _GNU_SOURCE?

Does it hurt to define it when it isn't needed?

**Comment by jszakmeister** [comment]
I’ve seen mixed results about that approach online, but thought the same
thing.

-John

**Comment by jamessan** [comment]
I've pulled out the `CMP0059` changes for now because that starts a big adventure into the land of replacing direct `CMAKE_C_FLAGS...` manipulation with generator expressions and `add_definitions()`/`add_compile_options()`.  I _do_ think it is possible to use those as a function replacement for what we're doing now, but we should decide whether we care about the fidelity of `nvim --version`'s `Compilation: ...` line first (c.f., https://github.com/neovim/neovim/pull/3701#issuecomment-157805550).

**Comment by justinmk** [comment]
> we should decide whether we care about the fidelity of `nvim --version`'s `Compilation: ...` line

We never cared about its fidelity, it has always been "best effort" or a "hint". Would be nice to keep it around to give some modicum of visibility. But if it's holding us back, let's remove it. We can restore it later if we find a way to get the build flags (is there a file in `build/` that we can scrape?).

**Review by justinmk — APPROVED** [review]

**Comment by jamessan** [comment]
> We never cared about its fidelity, it has always been "best effort" or a "hint".

Ok, that's good to know.  It's not "holding us back" (yet), so I'll leave it as is for now.  I at least have an idea of how to proceed when the time comes, so I'll get back to this once I've gotten past more important things on my TODO list.

Pending any other feedback, the PR is ready.

**Pull request merged**